[quant][pt2e] Support int16 quantization

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108453

Summary:
Previously we can only use native pytorch int dtypes that has corresponding quantized dtypes (e.g. quint8, qint8), this
PR removes this assumption in observers/fake_quants so that users can use all pytorch native dtypes (except for int64, we can add it later if need)
the main addition here is int16.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Support int16 quantization

Summary:
Previously we can only use native pytorch int dtypes that has corresponding quantized dtypes (e.g. quint8, qint8), this
PR removes this assumption in observers/fake_quants so that users can use all pytorch native dtypes (except for int64, we can add it later if need)
the main addition here is int16.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Support int16 quantization"


Summary:
Previously we can only use native pytorch int dtypes that has corresponding quantized dtypes (e.g. quint8, qint8), this
PR removes this assumption in observers/fake_quants so that users can use all pytorch native dtypes (except for int64, we can add it later if need)
the main addition here is int16.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/observer.py:
```diff
@@ -330,7 +338,7 @@ def _calculate_qparams(
             scale = max_val_pos / (float(quant_max - quant_min) / 2)
             scale = torch.max(scale, self.eps)
-            if self.dtype == torch.quint8:
+            if self.dtype in [torch.quint8, torch.uint8]:
```
why did this have to be changed?

jerryzh168 replied:
this is to make sure the code works for both quint8 and uint8 dtypes

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1277,6 +1281,66 @@ def validate(self, model: torch.fx.GraphModule) -> None:
+            def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+                # using int32 to simulate int16
+                int16_qspec = QuantizationSpec(
+                    dtype=torch.int32,
```
this says dtype=int32 not int16

kimishpatel replied:
oh nevermind, I see the restriction on the values. I thought int16 is supported torch dtype but not uint16? Then why not use that dtype directly

jerryzh168 replied:
oh I think I can use torch.int16 here directly, not sure why I used int32 in the beginning

kimishpatel reviewed (changes requested):
sending back for test fix.

For dtype change in fx workflow, previously it was working because we were mapping torch dtypes to quantized dtypes?

andrewor14 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -2114,7 +2177,7 @@ def _get_getitem_nodes(m: torch.fx.GraphModule):
     def test_qat_update_shared_qspec(self):
         """
-        Test the case where nodes used in SharedQuantizationSpec were replaced
+        Test the case where nodes used in SharedQuantizationpec were replaced
```
typo

jerryzh168 committed:
Update on "[quant][pt2e] Support int16 quantization"


Summary:
Previously we can only use native pytorch int dtypes that has corresponding quantized dtypes (e.g. quint8, qint8), this
PR removes this assumption in observers/fake_quants so that users can use all pytorch native dtypes (except for int64, we can add it later if need)
the main addition here is int16.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Support int16 quantization"


Summary:
Previously we can only use native pytorch int dtypes that has corresponding quantized dtypes (e.g. quint8, qint8), this
PR removes this assumption in observers/fake_quants so that users can use all pytorch native dtypes (except for int64, we can add it later if need)
the main addition here is int16.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> For dtype change in fx workflow, previously it was working because we were mapping torch dtypes to quantized dtypes?

yeah that's correct, we are not really using torch dtypes before in observer or fx flows

jerryzh168 committed:
Update on "[quant][pt2e] Support int16 quantization"


Summary:
Previously we can only use native pytorch int dtypes that has corresponding quantized dtypes (e.g. quint8, qint8), this
PR removes this assumption in observers/fake_quants so that users can use all pytorch native dtypes (except for int64, we can add it later if need)
the main addition here is int16.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
